archeryutils/targets.py

```diff
@@ -233,43 +285,84 @@ def min_score(self) -> float:
         float
             minimum score possible on this target face
 
-        Raises
-        ------
-        ValueError
-            If a scoring system is not accounted for in the function.
-
         Examples
         --------
         >>> mytarget = au.Target("10_zone", (122, "cm"), (70.0, "m"))
         >>> mytarget.min_score()
         1.0
         """
-        if self.scoring_system in (
-            "5_zone",
-            "10_zone",
-            "10_zone_compound",
-            "WA_field",
-            "IFAA_field_expert",
-            "Worcester",
-        ):
-            return 1.0
-        if self.scoring_system in (
-            "10_zone_6_ring",
-            "10_zone_6_ring_compound",
-        ):
-            return 5.0
-        if self.scoring_system in (
-            "10_zone_5_ring",
-            "10_zone_5_ring_compound",
-        ):
-            return 6.0
-        if self.scoring_system in ("Worcester_2_ring",):
-            return 4.0
-        if self.scoring_system in ("IFAA_field",):
-            return 3.0
-        if self.scoring_system in ("Beiter_hit_miss"):
-            # For Beiter options are hit and miss, so return 0 here
-            return 0.0
-        # NB: Should be hard (but not impossible) to get here without catching earlier.
-        msg = f"Target face '{self.scoring_system}' has no specified minimum score."
-        raise ValueError(msg)
+        return min(self.face_spec.values())
+
+    @property
+    def face_spec(self) -> FaceSpec:
+        """Get the targets face specification, generating on demand if needed."""
+        if self._face_spec is None:
+            if self.scoring_system == "Custom":
+                msg = (
+                    "Trying to generate face spec for custom target "
+                    "but no existing spec found: "
+                    "try instantiating with `Target.from_face_spec` instead"
+                )
+                raise ValueError(msg)
+            self._face_spec = self.gen_face_spec(self.scoring_system, self._diameter)
+        return MappingProxyType(self._face_spec)
```

Comment:
The indirection is needed somewhere, the issue is that `from_face_spec` calls `Target.__init__("Custom", *other_params`. And since as it stands the `face_spec` property special cases "Custom" and just looks up the attribute we get into problems because it hasn't been set yet.

I could special case handling "Custom" in `__init__` but that then felt like it was defeating the point of having a seperate constructor. 

Altneratively the `from_face_spec` classmethod could probably be done via `Target.__new__` instead of `__init__`, I can look to see if that makes things clearer.

Reply:
Ahhhh, OK.
Could `gen_face_spec` return `None` for `"custom"` and then have the setting of `_face_spec` in `from_face_spec`?

And I note as a result that the normal `__init__` will need to guard against `"Custom"` now being a recognised scoring system:
```python
>>> import archeryutils as au
>>> myTarget = au.Target("Custom", 20, 20)
>>> myTarget.face_spec()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/.../archeryutils/targets.py", line 306, in face_spec
    raise ValueError(msg)
ValueError: Trying to generate face spec for custom target but no existing spec found: try instantiating with `Target.from_face_spec` instead
```

Might be better not to have `"Custom"` as a 'recognised system' but instead to check:
```python
if scoring_system not in self.supported_systems + ("Custom", )
```
perhaps (could be nicer than adding tuples I expect)?

Reply:
`__new__` doesnt help here. We would still have to have a branch for "Custom" in `__init__` to skip attempting to generate the face_spec. It's not too bad and on consideration may be better than what was being done in the property before. It does mean that some of the tests that looked at force changing the scoring system to an invalid one no longer pass, because the target now happily returns the results that were correct at the time that it was initialised. But to be fair, the comments in the test say it well enough:

```python
        # Silence mypy as scoring_system must be a valid literal ScoringSystem
        # Ignore private attribute access to modify read only property
        # Conclusion: no way this happens by accident
        target._scoring_system = "InvalidScoringSystem"  # type: ignore[assignment]  # noqa: SLF001
```

The same issue would have already happened if someone changed to a different standard scoring system after accessing the face_spec so I am happy to just not worry about it.

Reply:
See what you think of this diff, I think it does work and is a bit clearer. Now the sanity check is done once at initialisation, and the property access isn't slowed down in the happy path.

```diff
❯ git diff
diff --git a/archeryutils/targets.py b/archeryutils/targets.py
index 9d139da..ca0d933 100644
--- a/archeryutils/targets.py
+++ b/archeryutils/targets.py
@@ -100,7 +100,7 @@ class Target:

     """

-    _face_spec: Optional[FaceSpec] = None
+    _face_spec: FaceSpec

     supported_systems = get_args(ScoringSystem)
     supported_distance_units = length.yard | length.metre
@@ -132,6 +132,9 @@ class Target:
         self._native_diameter_unit = native_diameter_unit
         self.indoor = indoor

+        if scoring_system != "Custom":
+            self._face_spec = self.gen_face_spec(scoring_system, self._diameter)
+
     @classmethod
     def from_face_spec(
         cls,
@@ -295,16 +298,19 @@ class Target:
     @property
     def face_spec(self) -> FaceSpec:
         """Get the targets face specification, generating on demand if needed."""
-        if self._face_spec is None:
-            if self.scoring_system == "Custom":
-                msg = (
-                    "Trying to generate face spec for custom target "
-                    "but no existing spec found: "
-                    "try instantiating with `Target.from_face_spec` instead"
-                )
-                raise ValueError(msg)
-            self._face_spec = self.gen_face_spec(self.scoring_system, self._diameter)
-        return MappingProxyType(self._face_spec)
+        # Still have some error handling in here for the case where
+        # users use the wrong initaliser:
+        # eg target = Target("Custom", 10, 10)
+        # As otherwise errors raised are somewhat cryptic
+        try:
+            return MappingProxyType(self._face_spec)
+        except AttributeError as err:
+            msg = (
+                "Trying to generate face spec for custom target "
+                "but no existing spec found: "
+                "try instantiating with `Target.from_face_spec` instead"
+            )
+            raise ValueError(msg) from err

     @staticmethod
     def gen_face_spec(system: ScoringSystem, diameter: float) -> FaceSpec:
```

Reply:
Yep, that looks good!